src/pybind/mgr/volumes/fs/operations/trash.py

```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
 
     def purge(self, trashpath, should_cancel):
         """
-        purge a trash entry.
+        Purge a trash entry with non-recursive depth-first approach.
+        Non-recursive aspect prevents hitting Python's recursion limit and
+        depth-first approach minimizes space complexity to avoid running out
+        of memory.
+
+        Besides, repetitive calls to fs.opendir() are prevented by storing
+        directory handle on stack and memory consumption is further reduced by
+        storing paths relative to trash path instead of absolute paths.
 
         :praram trash_entry: the trash entry to purge
         :praram should_cancel: callback to check if the purge should be aborted
         :return: None
         """
-        def rmtree(root_path):
-            log.debug("rmtree {0}".format(root_path))
-            try:
-                with self.fs.opendir(root_path) as dir_handle:
-                    d = self.fs.readdir(dir_handle)
-                    while d and not should_cancel():
-                        if d.d_name not in (b".", b".."):
-                            d_full = os.path.join(root_path, d.d_name)
-                            if d.is_dir():
-                                rmtree(d_full)
-                            else:
-                                self.fs.unlink(d_full)
-                        d = self.fs.readdir(dir_handle)
-            except cephfs.ObjectNotFound:
-                return
-            except cephfs.Error as e:
-                raise VolumeException(-e.args[0], e.args[1])
-            # remove the directory only if we were not asked to cancel
-            # (else we would fail to remove this anyway)
-            if not should_cancel():
-                self.fs.rmdir(root_path)
-
-        # catch any unlink errors
+        log.debug(f'purge(): trashpath = {trashpath}')
+
+        # XXX: add trailing '/' since its absence will produce relative path
+        # with leading '/' which will misguide os.path.join() to produce
+        # unexpected results.
+        #
+        # os.path.join('abcd', '123') = '/abcd/123'   --> expected
+        # os.path.join('abcd', '/123') = '/123'       --> unexpected
+        if trashpath.decode()[len(trashpath) - 1] != '/':
+            trashpath += b'/'
+
         try:
-            rmtree(trashpath)
+            # each stack member is tuple containing path to a directory and
+            # handle for that directory.
+            #
+            # The path here is a relative path (relative to trash path) instead
+            # of absolute path. This is to minimize the memory consumption and
+            # thereby also minimize the chance of running out of memory.
+            #
+            # The directory handle is also stored on stack to avoid unnecessary
+            # repetitive calls to fs.opendir().
+            stack = deque([(b'', self.fs.opendir(trashpath)), ])
```

Comment:
```suggestion
            stack = deque([(b'', self.fs.opendir(trashpath)), ])
```

a list is fine

Reply:
You mean a list instead of tuple? Or deque? Both tuple and deque are perform better in this scenario.